File: cli/src/specsfy_cli/updater.py
from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Any, TextIO

from . import __version__
from .github import api_headers


TAGS_API_URL = "https://api.github.com/repos/promovaweb/specsfy/tags?per_page=100"
UV_TOOL_NAME = "specsfy-cli"
SEMANTIC_TAG = re.compile(r"^v?(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)$")
COMMIT_SHA = re.compile(r"^[0-9a-f]{40,64}$")
DEFAULT_CHECK_INTERVAL_SECONDS = 86400
NETWORK_TIMEOUT_SECONDS = 4


@dataclass(frozen=True)
class UpdateInfo:
    current_version: str
    latest_version: str
    tag: str
    commit_sha: str


def global_config_path(*, home: Path | None = None) -> Path:
    root = home if home is not None else Path.home()
    return root / ".specsfy" / "cli.json"


def ensure_global_config(path: Path | None = None) -> dict[str, Any]:
    target = path or global_config_path()
    if target.is_file():
        payload = _read_global_config(target)
    else:
        payload = {}
    changed = False
    if payload.get("schema_version") != 1:
        payload["schema_version"] = 1
        changed = True
    settings = payload.setdefault("settings", {})
    if not isinstance(settings, dict):
        raise ValueError(f"settings inválido em {target}")
    defaults = {
        "check_updates_on_startup": True,
        "check_interval_seconds": DEFAULT_CHECK_INTERVAL_SECONDS,
    }
    for key, value in defaults.items():
        if key not in settings:
            settings[key] = value
            changed = True
    cache = payload.setdefault("cache", {})
    if not isinstance(cache, dict):
        raise ValueError(f"cache inválido em {target}")
    if changed or not target.is_file():
        _write_global_config(target, payload)
    else:
        target.chmod(0o600)
    return payload


def check_for_update(
    current_version: str = __version__,
    *,
    cache_path: Path | None = None,
    now: float | None = None,
    opener=None,
    force: bool = False,
) -> UpdateInfo | None:
    target = cache_path or global_config_path()
    payload = ensure_global_config(target)
    settings = payload["settings"]
    cache = payload["cache"]
    if not settings.get("check_updates_on_startup", True) and not force:
        return None

    checked_at = cache.get("last_checked_at")
    interval = settings.get(
        "check_interval_seconds",
        DEFAULT_CHECK_INTERVAL_SECONDS,
    )
    timestamp = time.time() if now is None else now
    if (
        not force
        and isinstance(checked_at, (int, float))
        and isinstance(interval, (int, float))
        and interval > 0
        and timestamp - checked_at < interval
    ):
        return _cached_update(current_version, cache)

    open_url = opener or urllib.request.urlopen
    headers = api_headers(f"specsfy-cli/{current_version}")
    etag = cache.get("etag")
    if isinstance(etag, str) and etag:
        headers["If-None-Match"] = etag
    request = urllib.request.Request(TAGS_API_URL, headers=headers)
    try:
        with open_url(request, timeout=NETWORK_TIMEOUT_SECONDS) as response:
            tags = json.loads(response.read().decode("utf-8"))
            response_etag = response.headers.get("ETag")
    except urllib.error.HTTPError as error:
        if error.code == 304:
            cache["last_checked_at"] = timestamp
            cache.pop("last_error", None)
            _write_global_config(target, payload)
            return _cached_update(current_version, cache)
        return _record_check_error(
            current_version,
            target,
            payload,
            timestamp,
            error,
        )
    except (OSError, UnicodeError, json.JSONDecodeError) as error:
        return _record_check_error(
            current_version,
            target,
            payload,
            timestamp,
            error,
        )

    try:
        latest = _latest_semantic_tag(tags)
    except ValueError as error:
        return _record_check_error(
            current_version,
            target,
            payload,
            timestamp,
            error,
        )
    cache["last_checked_at"] = timestamp
    cache.pop("last_error", None)
    if isinstance(response_etag, str):
        cache["etag"] = response_etag
    if latest is None:
        for key in ("latest_version", "latest_tag", "latest_commit"):
            cache.pop(key, None)
    else:
        version, tag, commit_sha = latest
        cache.update(
            {
                "latest_version": version,
                "latest_tag": tag,
                "latest_commit": commit_sha,
            }
        )
    _write_global_config(target, payload)
    return _cached_update(current_version, cache)


def uv_upgrade_command(uv_executable: str = "uv") -> list[str]:
    return [uv_executable, "tool", "upgrade", UV_TOOL_NAME]


def upgrade_with_uv(
    *,
    runner=subprocess.run,
    uv_executable: str | None = None,
) -> None:
    executable = uv_executable or shutil.which("uv")
    if executable is None:
        raise RuntimeError(
            "uv não foi encontrado no PATH; instale-o e execute "
            "`uv tool upgrade specsfy-cli`"
        )
    runner(uv_upgrade_command(executable), check=True)


def offer_startup_update(
    *,
    checker=check_for_update,
    upgrader=upgrade_with_uv,
    input_fn=input,
    output: TextIO = sys.stdout,
    interactive: bool | None = None,
) -> bool:
    is_interactive = (
        sys.stdin.isatty() and sys.stdout.isatty()
        if interactive is None
        else interactive
    )
    if not is_interactive:
        return False
    try:
        update = checker()
    except (ValueError, OSError) as error:
        print(f"Aviso: não foi possível verificar atualizações: {error}", file=output)
        return False
    if update is None:
        return False

    print(
        f"Uma nova versão do Specsfy CLI está disponível: "
        f"{update.current_version} → {update.latest_version}.",
        file=output,
    )
    answer = input_fn("Deseja atualizar agora? [s/N] ").strip().casefold()
    if answer not in {"s", "sim", "y", "yes"}:
        print("Atualização adiada. Abrindo a aplicação normalmente.", file=output)
        return False
    try:
        upgrader()
    except (RuntimeError, OSError, subprocess.SubprocessError) as error:
        print(f"Falha ao atualizar: {error}. Abrindo normalmente.", file=output)
        return False
    print(
        f"O uv atualizou o ambiente do Specsfy CLI para {update.latest_version}. "
        "O CLI será fechado; abra-o novamente para usar a nova versão.",
        file=output,
    )
    return True


def _read_global_config(path: Path) -> dict[str, Any]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as error:
        raise ValueError(f"configuração global inválida em {path}: {error.msg}") from error
    if not isinstance(payload, dict):
        raise ValueError(f"configuração global inválida em {path}")
    return payload


def _write_global_config(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    descriptor, temporary_name = tempfile.mkstemp(
        prefix=".cli-",
        suffix=".tmp",
        dir=path.parent,
    )
    temporary = Path(temporary_name)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8") as stream:
            json.dump(payload, stream, ensure_ascii=False, indent=2)
            stream.write("\n")
            stream.flush()
            os.fsync(stream.fileno())
        temporary.chmod(0o600)
        os.replace(temporary, path)
        path.chmod(0o600)
    finally:
        temporary.unlink(missing_ok=True)


def _latest_semantic_tag(tags: object) -> tuple[str, str, str] | None:
    if not isinstance(tags, list):
        raise ValueError("resposta de tags inválida")
    candidates: list[tuple[tuple[int, int, int], str, str]] = []
    for item in tags:
        if not isinstance(item, dict):
            continue
        tag = item.get("name")
        commit = item.get("commit")
        if not isinstance(tag, str) or not isinstance(commit, dict):
            continue
        match = SEMANTIC_TAG.fullmatch(tag)
        commit_sha = commit.get("sha")
        if (
            match is None
            or not isinstance(commit_sha, str)
            or COMMIT_SHA.fullmatch(commit_sha) is None
        ):
            continue
        version_tuple = tuple(
            int(match.group(name)) for name in ("major", "minor", "patch")
        )
        candidates.append((version_tuple, tag, commit_sha))
    if not candidates:
        return None
    version_tuple, tag, commit_sha = max(candidates)
    version = ".".join(str(part) for part in version_tuple)
    return version, tag, commit_sha


def _version_tuple(version: str) -> tuple[int, int, int]:
    match = SEMANTIC_TAG.fullmatch(version)
    if match is None:
        raise ValueError(f"versão semântica inválida: {version}")
    return tuple(int(match.group(name)) for name in ("major", "minor", "patch"))


def _cached_update(
    current_version: str,
    cache: dict[str, Any],
) -> UpdateInfo | None:
    latest_version = cache.get("latest_version")
    tag = cache.get("latest_tag")
    commit_sha = cache.get("latest_commit")
    if not all(isinstance(value, str) for value in (latest_version, tag, commit_sha)):
        return None
    if COMMIT_SHA.fullmatch(commit_sha) is None:
        return None
    tag_match = SEMANTIC_TAG.fullmatch(tag)
    if tag_match is None or _version_tuple(tag) != _version_tuple(latest_version):
        return None
    if _version_tuple(latest_version) <= _version_tuple(current_version):
        return None
    return UpdateInfo(
        current_version=current_version,
        latest_version=latest_version,
        tag=tag,
        commit_sha=commit_sha,
    )


def _record_check_error(
    current_version: str,
    path: Path,
    payload: dict[str, Any],
    timestamp: float,
    error: Exception,
) -> UpdateInfo | None:
    cache = payload["cache"]
    cache["last_checked_at"] = timestamp
    cache["last_error"] = str(error)
    _write_global_config(path, payload)
    return _cached_update(current_version, cache)

File: cli/src/updater.ts
/**
 * Verificação consentida de versões e atualização do pacote ou executável.
 */

import { execFile } from "node:child_process";
import { chmod, mkdir, rename, rm, writeFile } from "node:fs/promises";
import { homedir } from "node:os";
import { basename, dirname, extname, join } from "node:path";
import { promisify } from "node:util";
import { apiHeaders } from "./github.js";
import { isFile, readJson, writeTextAtomic } from "./filesystem.js";

const execFileAsync = promisify(execFile);

export const TAGS_API_URL =
  "https://api.github.com/repos/promovaweb/specsfy/tags?per_page=100";
export const NPM_REGISTRY_URL =
  "https://registry.npmjs.org/@promovaweb%2fspecsfy/latest";
export const EXECUTABLE_DOWNLOAD_URL = "https://get.specsfy.dev";
export const NPM_PACKAGE_NAME = "@promovaweb/specsfy";
export const DEFAULT_CHECK_INTERVAL_SECONDS = 86_400;
const EXECUTABLE_DOWNLOAD_TIMEOUT_MS = 15_000;
const SEMANTIC_TAG = /^v?(\d+)\.(\d+)\.(\d+)$/;
const COMMIT_SHA = /^[0-9a-f]{40,64}$/;

/** Versão estável mais recente que supera a versão em execução. */
export interface UpdateInfo {
  current_version: string;
  latest_version: string;
  tag: string;
  commit_sha?: string;
}

interface GlobalConfig {
  schema_version: 1;
  settings: Record<string, unknown> & {
    check_updates_on_startup: boolean;
    check_interval_seconds: number;
  };
  cache: Record<string, unknown>;
  [key: string]: unknown;
}

export type UpgradeTarget = "npm" | "executable";

/** Retorna o caminho global do cache do CLI. */
export function globalConfigPath(home = homedir()): string {
  return join(home, ".specsfy", "cli.json");
}

/** Cria ou normaliza o cache global com permissão privada. */
export async function ensureGlobalConfig(
  path = globalConfigPath(),
): Promise<GlobalConfig> {
  const raw = (await isFile(path)) ? await readJson(path) : {};
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error(`configuração global inválida em ${path}`);
  }
  const payload = raw as Record<string, unknown>;
  payload.schema_version = 1;
  if (
    payload.settings !== undefined &&
    (!payload.settings ||
      typeof payload.settings !== "object" ||
      Array.isArray(payload.settings))
  ) {
    throw new Error(`settings inválido em ${path}`);
  }
  if (
    payload.cache !== undefined &&
    (!payload.cache ||
      typeof payload.cache !== "object" ||
      Array.isArray(payload.cache))
  ) {
    throw new Error(`cache inválido em ${path}`);
  }
  const settings = (payload.settings as Record<string, unknown> | undefined) ?? {};
  const cache = (payload.cache as Record<string, unknown> | undefined) ?? {};
  settings.check_updates_on_startup ??= true;
  settings.check_interval_seconds ??= DEFAULT_CHECK_INTERVAL_SECONDS;
  payload.settings = settings;
  payload.cache = cache;
  await writeGlobalConfig(path, payload as GlobalConfig);
  return payload as GlobalConfig;
}

/** Consulta a versão publicada e retorna somente atualização distribuída. */
export async function checkForUpdate(
  currentVersion: string,
  options: {
    cachePath?: string;
    now?: number;
    force?: boolean;
    fetcher?: typeof fetch;
  } = {},
): Promise<UpdateInfo | undefined> {
  const target = options.cachePath ?? globalConfigPath();
  const payload = await ensureGlobalConfig(target);
  const { settings, cache } = payload;
  if (settings.check_updates_on_startup === false && !options.force) return undefined;
  const timestamp = options.now ?? Date.now() / 1000;
  const checkedAt = cache.last_checked_at;
  const interval = settings.check_interval_seconds;
  if (
    !options.force &&
    typeof checkedAt === "number" &&
    typeof interval === "number" &&
    interval > 0 &&
    timestamp - checkedAt < interval &&
    typeof cache.latest_distribution_version === "string"
  ) {
    return cachedUpdate(currentVersion, cache);
  }
  const fetcher = options.fetcher ?? fetch;
  try {
    const registryHeaders: Record<string, string> = {
      Accept: "application/json",
      "User-Agent": `specsfy-cli/${currentVersion}`,
    };
    if (typeof cache.registry_etag === "string" && cache.registry_etag) {
      registryHeaders["If-None-Match"] = cache.registry_etag;
    }
    const registryResponse = await fetcher(NPM_REGISTRY_URL, {
      headers: registryHeaders,
      signal: AbortSignal.timeout(4_000),
    });
    let publishedVersion: string;
    if (registryResponse.status === 304) {
      const cachedVersion = cache.latest_distribution_version;
      if (typeof cachedVersion !== "string") {
        throw new Error("cache do registro npm sem versão publicada");
      }
      publishedVersion = cachedVersion;
    } else {
      if (!registryResponse.ok) throw new Error(`HTTP ${registryResponse.status}`);
      publishedVersion = latestPublishedVersion(
        (await registryResponse.json()) as unknown,
      );
      const registryEtag = registryResponse.headers.get("etag");
      if (registryEtag) cache.registry_etag = registryEtag;
    }

    let taggedVersion: ReturnType<typeof latestSemanticTag>;
    try {
      const headers = await apiHeaders(`specsfy-cli/${currentVersion}`);
      if (typeof cache.etag === "string" && cache.etag) {
        headers["If-None-Match"] = cache.etag;
      }
      const tagResponse = await fetcher(TAGS_API_URL, {
        headers,
        signal: AbortSignal.timeout(4_000),
      });
      if (tagResponse.status === 304) {
        taggedVersion = cachedTag(cache);
      } else if (tagResponse.ok) {
        taggedVersion = latestSemanticTag((await tagResponse.json()) as unknown);
        const tagEtag = tagResponse.headers.get("etag");
        if (tagEtag) cache.etag = tagEtag;
      }
    } catch {
      taggedVersion = cachedTag(cache);
    }

    cache.last_checked_at = timestamp;
    delete cache.last_error;
    cache.latest_distribution_version = publishedVersion;
    cache.latest_version = publishedVersion;
    cache.latest_tag =
      taggedVersion?.version === publishedVersion
        ? taggedVersion.tag
        : `v${publishedVersion}`;
    if (taggedVersion?.version === publishedVersion) {
      cache.latest_commit = taggedVersion.commit;
    } else {
      delete cache.latest_commit;
    }
    await writeGlobalConfig(target, payload);
    return cachedUpdate(currentVersion, cache);
  } catch (error) {
    cache.last_checked_at = timestamp;
    cache.last_error = error instanceof Error ? error.message : String(error);
    await writeGlobalConfig(target, payload);
    return cachedUpdate(currentVersion, cache);
  }
}

/** Comando público usado para atualizar a instalação global. */
export function npmUpgradeArguments(): string[] {
  return ["install", "--global", `${NPM_PACKAGE_NAME}@latest`];
}

/** Atualiza o pacote pelo npm encontrado no PATH. */
export async function upgradeWithNpm(): Promise<void> {
  await execFileAsync("npm", npmUpgradeArguments(), { encoding: "utf8" });
}

/** Retorna a entrada quando o CLI está rodando como executável avulso. */
export function standaloneExecutablePath(
  argv = process.argv,
): string | undefined {
  const entry = argv[1];
  if (!entry || extname(entry)) return undefined;
  return entry;
}

/** Baixa, valida e substitui atomicamente o executável avulso em uso. */
export async function upgradeWithExecutable(
  executablePath: string,
  expectedVersion: string,
  fetcher: typeof fetch = fetch,
): Promise<void> {
  const response = await fetcher(EXECUTABLE_DOWNLOAD_URL, {
    headers: { "User-Agent": `specsfy-cli/${expectedVersion}` },
    signal: AbortSignal.timeout(EXECUTABLE_DOWNLOAD_TIMEOUT_MS),
  });
  if (!response.ok) throw new Error(`HTTP ${response.status}`);

  const temporary = join(
    dirname(executablePath),
    `.${basename(executablePath)}.${process.pid}.tmp`,
  );
  try {
    await writeFile(temporary, Buffer.from(await response.arrayBuffer()), {
      mode: 0o755,
    });
    await chmod(temporary, 0o755);
    const result = await execFileAsync(temporary, ["--version"], {
      encoding: "utf8",
    });
    if (result.stdout.trim() !== expectedVersion) {
      throw new Error(
        `executável baixado informa ${result.stdout.trim() || "versão vazia"}; esperado ${expectedVersion}`,
      );
    }
    await rename(temporary, executablePath);
    await chmod(executablePath, 0o755);
  } finally {
    await rm(temporary, { force: true });
  }
}

/** Atualiza a instalação real detectada no processo atual. */
export async function upgradeCurrentInstallation(
  expectedVersion: string,
  options: {
    executablePath?: string;
    fetcher?: typeof fetch;
  } = {},
): Promise<UpgradeTarget> {
  const executablePath = options.executablePath ?? standaloneExecutablePath();
  if (executablePath) {
    await upgradeWithExecutable(
      executablePath,
      expectedVersion,
      options.fetcher ?? fetch,
    );
    return "executable";
  }
  await upgradeWithNpm();
  return "npm";
}

/** Conduz o prompt de consentimento e informa se o processo deve encerrar. */
export async function offerStartupUpdate(
  currentVersion: string,
  ask: (prompt: string) => Promise<string>,
  output: (line: string) => void = console.log,
): Promise<boolean> {
  if (!process.stdin.isTTY || !process.stdout.isTTY) return false;
  let update: UpdateInfo | undefined;
  try {
    update = await checkForUpdate(currentVersion);
  } catch (error) {
    output(
      `Aviso: não foi possível verificar atualizações: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
    return false;
  }
  if (!update) return false;
  output(
    `Uma nova versão do Specsfy CLI está disponível: ` +
      `${update.current_version} → ${update.latest_version}.`,
  );
  const answer = (await ask("Deseja atualizar agora? [s/N] ")).trim().toLowerCase();
  if (!["s", "sim", "y", "yes"].includes(answer)) {
    output("Atualização adiada. Abrindo a aplicação normalmente.");
    return false;
  }
  let target: UpgradeTarget;
  try {
    target = await upgradeCurrentInstallation(update.latest_version);
  } catch (error) {
    output(
      `Falha ao atualizar: ${
        error instanceof Error ? error.message : String(error)
      }. Abrindo normalmente.`,
    );
    return false;
  }
  output(
    `${target === "executable" ? "O executável" : "O npm"} atualizou o Specsfy CLI para ${update.latest_version}. ` +
      "O CLI será fechado; abra-o novamente para usar a nova versão.",
  );
  return true;
}

async function writeGlobalConfig(
  path: string,
  payload: GlobalConfig,
): Promise<void> {
  const directory = dirname(path);
  await mkdir(directory, { recursive: true, mode: 0o700 });
  if (basename(directory) === ".specsfy") await chmod(directory, 0o700);
  await writeTextAtomic(path, `${JSON.stringify(payload, null, 2)}\n`, 0o600);
  await chmod(path, 0o600);
}

function latestSemanticTag(
  tags: unknown,
): { version: string; tag: string; commit: string } | undefined {
  if (!Array.isArray(tags)) throw new Error("resposta de tags inválida");
  const candidates: Array<{
    tuple: [number, number, number];
    version: string;
    tag: string;
    commit: string;
  }> = [];
  for (const item of tags) {
    if (!item || typeof item !== "object") continue;
    const value = item as Record<string, unknown>;
    const match = typeof value.name === "string" ? SEMANTIC_TAG.exec(value.name) : null;
    const commit =
      value.commit && typeof value.commit === "object"
        ? (value.commit as Record<string, unknown>).sha
        : undefined;
    if (!match || typeof commit !== "string" || !COMMIT_SHA.test(commit)) continue;
    const tuple = [
      Number(match[1]),
      Number(match[2]),
      Number(match[3]),
    ] as [number, number, number];
    candidates.push({
      tuple,
      version: tuple.join("."),
      tag: value.name as string,
      commit,
    });
  }
  candidates.sort((left, right) => compareVersions(right.tuple, left.tuple));
  return candidates[0];
}

/** Extrai a versão estável realmente publicada no registro npm. */
function latestPublishedVersion(payload: unknown): string {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new Error("resposta de versão publicada inválida");
  }
  const version = (payload as Record<string, unknown>).version;
  if (typeof version !== "string" || !SEMANTIC_TAG.test(version)) {
    throw new Error("versão publicada inválida");
  }
  return versionTuple(version).join(".");
}

/** Recupera a tag em cache quando a API do GitHub responde com 304. */
function cachedTag(
  cache: Record<string, unknown>,
): { version: string; tag: string; commit: string } | undefined {
  const version = cache.latest_version;
  const tag = cache.latest_tag;
  const commit = cache.latest_commit;
  if (
    typeof version !== "string" ||
    typeof tag !== "string" ||
    typeof commit !== "string" ||
    !COMMIT_SHA.test(commit)
  ) {
    return undefined;
  }
  return { version, tag, commit };
}

function cachedUpdate(
  currentVersion: string,
  cache: Record<string, unknown>,
): UpdateInfo | undefined {
  const latest = cache.latest_version;
  const tag = cache.latest_tag;
  const commit = cache.latest_commit;
  if (
    typeof latest !== "string" ||
    typeof tag !== "string" ||
    cache.latest_distribution_version !== latest
  ) {
    return undefined;
  }
  const currentTuple = versionTuple(currentVersion);
  const latestTuple = versionTuple(latest);
  const tagTuple = versionTuple(tag);
  if (
    compareVersions(latestTuple, currentTuple) <= 0 ||
    compareVersions(tagTuple, latestTuple) !== 0
  ) {
    return undefined;
  }
  const update: UpdateInfo = {
    current_version: currentVersion,
    latest_version: latest,
    tag,
  };
  if (typeof commit === "string" && COMMIT_SHA.test(commit)) {
    update.commit_sha = commit;
  }
  return update;
}

function versionTuple(version: string): [number, number, number] {
  const match = SEMANTIC_TAG.exec(version);
  if (!match) throw new Error(`versão semântica inválida: ${version}`);
  return [Number(match[1]), Number(match[2]), Number(match[3])];
}

function compareVersions(
  left: [number, number, number],
  right: [number, number, number],
): number {
  for (let index = 0; index < 3; index += 1) {
    const difference = left[index]! - right[index]!;
    if (difference) return difference;
  }
  return 0;
}
